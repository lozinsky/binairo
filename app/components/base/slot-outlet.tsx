import { type ReactNode, createContext, useContext } from 'react';

export const SlotOutletContext = createContext<ReactNode>(null);

export function SlotOutlet() {
  return useContext(SlotOutletContext);
}
