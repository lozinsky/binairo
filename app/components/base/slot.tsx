import { type HTMLAttributes, cloneElement, isValidElement } from 'react';

import { assert } from '~/shared/assert';

export function Slot({ children, ...props }: HTMLAttributes<HTMLElement>) {
  assert(isValidElement<HTMLAttributes<HTMLElement>>(children));

  return cloneElement(children, props);
}
