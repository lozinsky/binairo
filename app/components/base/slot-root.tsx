import { type ReactNode } from 'react';

import { SlotOutletContext } from '~/components/base/slot-outlet';

export function SlotRoot({ children, target }: { children: ReactNode; target: ReactNode }) {
  return <SlotOutletContext.Provider value={target}>{children}</SlotOutletContext.Provider>;
}
