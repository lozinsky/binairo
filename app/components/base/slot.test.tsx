/* eslint-disable react/jsx-key */
/* @vitest-environment jsdom */

import { render } from '@testing-library/react';
import { createRef, type ElementRef, forwardRef, useImperativeHandle, useRef } from 'react';
import { expect, test } from 'vitest';

import { Slot } from './slot';
import { SlotOutlet } from './slot-outlet';
import { SlotRoot } from './slot-root';

test.each([
  <Slot>
    <span>
      Foo
      <span>Bar</span>
    </span>
  </Slot>,
  <Slot className='slot'>
    <span>
      Foo
      <span>Bar</span>
    </span>
  </Slot>,
  <Slot className='slot'>
    <span className='foo'>
      Foo
      <span>Bar</span>
    </span>
  </Slot>,
  <Slot data-slot='slot'>
    <span>
      Foo
      <span>Bar</span>
    </span>
  </Slot>,
  <Slot data-slot='slot'>
    <span data-foo='foo'>
      Foo
      <span>Bar</span>
    </span>
  </Slot>,
  <Slot>
    <SlotRoot target={<span>Foo</span>}>
      <SlotOutlet />
      <span>Bar</span>
    </SlotRoot>
  </Slot>,
  <Slot>
    <SlotRoot target={<span>Bar</span>}>
      Foo
      <span>
        <SlotOutlet />
      </span>
    </SlotRoot>
  </Slot>,
  <Slot className='slot'>
    <SlotRoot target={<span>Foo</span>}>
      <SlotOutlet />
      <span>Bar</span>
    </SlotRoot>
  </Slot>,
  <Slot className='slot'>
    <SlotRoot target={<span className='foo'>Foo</span>}>
      <SlotOutlet />
      <span>Bar</span>
    </SlotRoot>
  </Slot>,
  <Slot data-slot='slot'>
    <SlotRoot target={<span>Foo</span>}>
      <SlotOutlet />
      <span>Bar</span>
    </SlotRoot>
  </Slot>,
  <Slot data-slot='slot'>
    <SlotRoot target={<span data-foo='foo'>Foo</span>}>
      <SlotOutlet />
      <span>Bar</span>
    </SlotRoot>
  </Slot>,
])('renders slot', (children) => {
  expect(render(children).asFragment()).toMatchSnapshot();
});

test.each([
  forwardRef<unknown>(function Component(props, ref) {
    const slotRef = useRef<ElementRef<typeof Slot>>(null);

    useImperativeHandle(ref, () => ({
      slotRef,
    }));

    return (
      <Slot ref={slotRef}>
        <span>
          Foo
          <span>Bar</span>
        </span>
      </Slot>
    );
  }),
  forwardRef<unknown>(function Component(props, ref) {
    const slotRef = useRef<ElementRef<typeof Slot>>(null);

    useImperativeHandle(ref, () => ({
      slotRef,
    }));

    return (
      <Slot ref={slotRef}>
        <SlotRoot target={<span>Foo</span>}>
          <SlotOutlet />
          <span>Bar</span>
        </SlotRoot>
      </Slot>
    );
  }),
  forwardRef<unknown>(function Component(props, ref) {
    const slotRef = useRef<ElementRef<typeof Slot>>(null);
    const spanRef = useRef<ElementRef<'span'>>(null);

    useImperativeHandle(ref, () => ({
      slotRef,
      spanRef,
    }));

    return (
      <Slot ref={slotRef}>
        <span ref={spanRef}>
          Foo
          <span>Bar</span>
        </span>
      </Slot>
    );
  }),
  forwardRef<unknown>(function Component(props, ref) {
    const slotRef = useRef<ElementRef<typeof Slot>>(null);
    const spanRef = useRef<ElementRef<'span'>>(null);

    useImperativeHandle(ref, () => ({
      slotRef,
      spanRef,
    }));

    return (
      <Slot ref={slotRef}>
        <SlotRoot target={<span ref={spanRef}>Foo</span>}>
          <SlotOutlet />
          <span>Bar</span>
        </SlotRoot>
      </Slot>
    );
  }),
])('renders slot with ref', (Component) => {
  const ref = createRef();

  render(<Component ref={ref} />);
  expect(ref.current).toMatchSnapshot();
});
