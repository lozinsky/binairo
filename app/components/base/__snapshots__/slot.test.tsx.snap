// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders slot 1`] = `
<DocumentFragment>
  <span>
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 2`] = `
<DocumentFragment>
  <span
    class="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 3`] = `
<DocumentFragment>
  <span
    class="slot foo"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 4`] = `
<DocumentFragment>
  <span
    data-slot="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 5`] = `
<DocumentFragment>
  <span
    data-foo="foo"
    data-slot="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 6`] = `
<DocumentFragment>
  <span>
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 7`] = `
<DocumentFragment>
  <span>
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 8`] = `
<DocumentFragment>
  <span
    class="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 9`] = `
<DocumentFragment>
  <span
    class="slot foo"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 10`] = `
<DocumentFragment>
  <span
    data-slot="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot 11`] = `
<DocumentFragment>
  <span
    data-foo="foo"
    data-slot="slot"
  >
    Foo
    <span>
      Bar
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders slot with ref 1`] = `
{
  "slotRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
}
`;

exports[`renders slot with ref 2`] = `
{
  "slotRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
}
`;

exports[`renders slot with ref 3`] = `
{
  "slotRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
  "spanRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
}
`;

exports[`renders slot with ref 4`] = `
{
  "slotRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
  "spanRef": {
    "current": <span>
      Foo
      <span>
        Bar
      </span>
    </span>,
  },
}
`;
